### ISSUE TITLE
Proposal to Add Identity

### ISSUE THREAD

**Issue by vishvananda** [twelve-factor/twelve-factor#9]
### Rationale

To support Backing services, an app needs to be properly authenticated by the backing service. While this can be done via credentials provided by the backing service, the most secure way to handle this is for the app to have a short-lived credential associated with its identity which the backing service can validate. Because the credential is-short lived, its value changes frequently, so it should be stored in a way that:
* is automatically provided by the platform
* is available only at runtime (not used during builds)
* notifies the app when the credential is updated

A single credential can introduce the “confused deputy problem”, where a receiving service can misuse the credential and impersonate the app with a third-party. The platform therefore needs to provide a credential that is scoped to a single backing service. Ideally, this credential would be associated with the backing service connection, so the lifecycle of the credential can be managed along with the connection.

### Why This Shouldn't Be Included in Another Factor?

We could explicitly cover identity in config, but this expands the scope of config significantly. Keeping identity as its own factor allows us to include the important things to know about identity.

### Open Questions

1. Does the identity credential depend on a new type of config (dynamic config), or does it explicitly use its own mechanism that is separate from config?
2. Which identity types should we give examples of?
* x509-svid
* jwt-svid
* oidc

**Comment by dstengle** [comment]
I think identity and authorization need to be independent of config. The bootstrapping and maintenance of identity might be config related, but the appropriate level of identity validation and authorization of actions goes far beyond config. This features largely in zero-trust, something that should figure into the new factors.

Config is a supporting domain and should be scoped pretty narrowly to avoid confusion with other cloud native behaviors.

**Comment by jesperfj** [comment]
I missed this issue. Just posted this loose thought in Discord:

> Imagine we came up with a standard location for "token files". A token file is a file that an app must watch for changes and upon change, load the new content (or something like that). We could specify a file and network location to cover both options. This behavior would be fairly trivial to implement in a small helper library in most programming languages and it could form a foundation underneath more vendor specific libraries such as aws-sdk.

I haven't thought through the confused deputy problem that @vishvananda raises regarding a single workload identity. But to me it almost doesn't matter because we have an opportunity to come up with something that can work for one as well as many tokens and that gives much more flexibility. A token that represents a single identity relies on an IAM system to handle authorization. A mechanism that supports any amount of tokens makes such an IAM system optional.

**Comment by jesperfj** [comment]
For both files and network locations, there are standard ways of being notified of change which means we don't have to invent a notification mechanism (I think).

**Comment by vishvananda** [comment]
Yes I'm a big fan of the filesystem + notification option. I think the advantage of a file is we can depend on fsnotify. Network based identity  has issues around signalling updates:
1. There is a bootstrapping problem. If the other end is actually on the network, you need a secure identity to talk to the network service. You could work around this with a local isolated network variant (similar to how the aws metadata api works) but this feels more complicated to implement locally
2. There is no clear way to notify without keeping a connection to the network service open, which seems brittle. If we want to support network identities as well then I think we'd want to settle on a posix signal approach for changes. Perhaps the app could get sent SIGPOLL or SIGUSR1 when an identity file changes. Unfortunately this doesn't allow us to differentiate if there are multiple identities.

Another alternative is that client code could be checking expiration of the token and explicitly reload it when the token is close to expiry. This could introduce its own set of errors though.

**Comment by jesperfj** [comment]
Coming up with an option for both file system and network will provide important flexibility for the various implementations. I can imagine many platforms not having an easy way to write to a containerized (or otherwise isolated) file system. E.g. I think the ec2 architecture was what prompted AWS to use a network endpoint (but I am guessing).

Yes, the network endpoint must itself only be accessible to the application process so we don't need another token for that.

With respect to notification... I wonder if we can get away with an approach that involves calling a function as opposed to accessing a variable (in whatever language this is implemented in).

If we think developers has a strong desire to access a token value in a (global) variable (similar to an env var), then for the file system approach with fsnotify, you'd need to boot up a separate thread that watches the file and updates the variable. That's not easy in some languages.

If we think developers will be ok with calling a function, then things get easier because now you can do some kind of read-through cache or even non-cached dynamic reads from either file or network socket. If our spec includes a way to specific ttl for a token value, then it'll be simple to implement caches for high throughput scenarios.

Think about something like dotenv. It gradually spread to all languages and developers now generally expect there to be some kind of dotenv library they can use to load values from a .env file with various options for setting them as "synthetic" env vars or otherwise be available in the program. It would be pretty cool if we could achieve a similar kind of dispersal and adoption of this.

**Comment by vishvananda** [comment]
@jesperfj if the application is calling a function to retrieve a token over the network, is it the application/frameworks job to check for expiry? Or is it just supposed to keep using the token until it fails and then request a new one on failure? The ergonomics of both of those approaches seem strictly worse to me and i would argue for preferring the filesystem approach. If we think flexibility in implementation is more important then we need to be explicit about how refresh/rotation is supposed to work so that authors can implement it consistently.

**Comment by jesperfj** [comment]
You can ask the same question for the filesystem approach right? Who is responsible for ensuring the token hasn't expired? That's an important question in either case, isn't it?

**Comment by halorgium** [comment]
To some extent, one can normalize both file, socket, network into a standard URI. The critical part is that the first hop to the token source or vendor does not require additional identification to bootstrap. 

I can imagine a world in which a hypervisor might vend tokens dynamically for local and trusted parties, based on the source process owner or based on the source network address. 

```
export WORKLOAD_ID_SOURCE=http://169.254.169.254/workload-id/svid-jwt
```

**Comment by jesperfj** [comment]
That makes sense.

We would still have to discuss the semantics of different sources as @vishvananda pointed out.

As far as I can tell, it is always the app code's responsibility to re-read from the source. In the case of a file, the app can use the standard OS notify syscall, but the app code still needs to listen for this notification, e.g. in a go routine, separate thread or as part of the global event loop.

In the case of an HTTP URI, standard cache headers can be used to instruct the app code when to re-read. This re-read can either happen similar to above or it can happen "on demand" by wrapping the value in a function call, basically a small internal read-through cache. This read-through cache mechanism can also be used in the case of a file by checking the modified timestamp instead of relying on the notify syscall.

So in both cases, we can declare (in 12factor) that these options should be made available by the control plane and if properly used in the app code, the app will not experience failures due to token expiration.

We could also say (in 12factor) that the factor is about dynamic reading of token material and 12factor is not opinionated about these other things. In some cases, the control plane might not indicate the need to refresh and the app is on its own and must be able to recover from failed requests by refreshing the token and retrying. I implemented this in https://github.com/jesperfj/force-rest-api.
